File: ruby/lib/woothee/misc.rb
# -*- coding: utf-8 -*-

require 'woothee/dataset'
require 'woothee/util'

module Woothee::Misc
  extend Woothee::Util

  def self.challenge_desktoptools(ua, result)
    data = case
           when ua.index('AppleSyndication/') then Woothee::DataSet.get('SafariRSSReader')
           when ua.index('compatible; Google Desktop/') then Woothee::DataSet.get('GoogleDesktop')
           when ua.index('Windows-RSS-Platform') then Woothee::DataSet.get('WindowsRSSReader')
           else nil
           end
    return false unless data

    update_map(result, data)
    true
  end

  def self.challenge_smartphone_patterns(ua, result)
    if ua.index('CFNetwork/')
      data = Woothee::DataSet.get('iOS')
      update_category(result, data[Woothee::KEY_CATEGORY])
      update_os(result, data[Woothee::KEY_NAME])
      return true
    end

    false
  end

  def self.challenge_http_library(ua, result)
    data,version = case
                   when ua =~ /^(?:Apache-HttpClient\/|Jakarta Commons-HttpClient\/|Java\/)/o || ua =~ /[- ]HttpClient(\/|$)/o
                     [Woothee::DataSet.get('HTTPLibrary'), 'Java']
                   when ua.index('Java(TM) 2 Runtime Environment,')
                     [Woothee::DataSet.get('HTTPLibrary'), 'Java']
                   when ua =~ /^Wget/o
                     [Woothee::DataSet.get('HTTPLibrary'), 'wget']
                   when ua =~ /^(?:libwww-perl|WWW-Mechanize|LWP::Simple|LWP |lwp-trivial)/o
                     [Woothee::DataSet.get('HTTPLibrary'), 'perl']
                   when ua =~ /^(?:Ruby|feedzirra|Typhoeus)/o
                     [Woothee::DataSet.get('HTTPLibrary'), 'ruby']
                   when ua =~ /^(Python-urllib\/|Twisted )/o
                     [Woothee::DataSet.get('HTTPLibrary'), 'python']
                   when ua =~ /^(?:PHP|WordPress|CakePHP|PukiWiki|PECL::HTTP)(?:\/| |$)/o || ua =~ /(?:PEAR |)HTTP_Request(?: class|2)/o
                     [Woothee::DataSet.get('HTTPLibrary'), 'php']
                   else [nil,nil]
                   end
    return false unless data

    update_map(result, data)
    update_version(result, version)
    true
  end

  def self.challenge_maybe_rss_reader(ua, result)
    data = if ua =~ /rss(?:reader|bar|[-_ \/;()]|[ +]*\/)/oi || ua =~ /headline-reader/oi
             Woothee::DataSet.get('VariousRSSReader')
           elsif ua.index('cococ/')
             Woothee::DataSet.get('VariousRSSReader')
           else
             nil
           end
    return false unless data

    update_map(result, data)
    true
  end
end

File: python/lib/woothee/misc.py
# -*- coding: utf-8 -*-

import re
from . import dataset
from . import util

def challenge_desktoptools(ua, result):
  if 'AppleSyndication/' in ua:
    data = dataset.get('SafariRSSReader')
  elif 'compatible; Google Desktop/' in ua:
    data = dataset.get('GoogleDesktop')
  elif 'Windows-RSS-Platform' in ua:
    data = dataset.get('WindowsRSSReader')
  else:
    return False
  util.update_map(result, data)
  return True

def challenge_smartphone_patterns(ua, result):
  if 'CFNetwork/' in ua:
    data = dataset.get('iOS')
    util.update_category(result, data[dataset.KEY_CATEGORY])
    util.update_os(result, data[dataset.KEY_NAME])
    return True
  return False

def challenge_http_library(ua, result):
  if re.search('^(?:Apache-HttpClient/|Jakarta Commons-HttpClient/|Java/)', ua) or re.search('[- ]HttpClient(/|$)', ua):
    data, version = dataset.get('HTTPLibrary'), 'Java'
  elif 'Java(TM) 2 Runtime Environment,' in ua:
    data, version = dataset.get('HTTPLibrary'), 'Java'
  elif re.search('^Wget', ua):
    data, version = dataset.get('HTTPLibrary'), 'wget'
  elif re.search('^(?:libwww-perl|WWW-Mechanize|LWP::Simple|LWP |lwp-trivial)', ua):
    data, version = dataset.get('HTTPLibrary'), 'perl'
  elif re.search('^(?:Ruby|feedzirra|Typhoeus)', ua):
    data, version = dataset.get('HTTPLibrary'), 'ruby'
  elif re.search('^(?:Python-urllib/|Twisted )', ua):
    data, version = dataset.get('HTTPLibrary'), 'python'
  elif re.search('^(?:PHP|WordPress|CakePHP|PukiWiki|PECL::HTTP)(?:/| |$)', ua) or re.search('(?:PEAR |)HTTP_Request(?: class|2)', ua):
    data, version = dataset.get('HTTPLibrary'), 'php'
  else:
    return False
  util.update_map(result, data)
  util.update_version(result, version)
  return True

def challenge_maybe_rss_reader(ua, result):
  if re.search('rss(?:reader|bar|[-_ /;()]|[ +]*/)', ua, re.I) or re.search('headline-reader', ua, re.I):
    data = dataset.get('VariousRSSReader')
  elif 'cococ/' in ua:
    data = dataset.get('VariousRSSReader')
  else:
    return False
  util.update_map(result, data)
  return True
